### PR TITLE
feat(idempotency): support dataclasses & pydantic models payloads

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:**

## Description of changes:

Add support for serialization of python dataclasses, but checking if `__dataclass_fields__` exists and calling `dataclasses.asdict`. This happens early on in the idempotency work flow.

### Dataclasses support

```python
from dataclasses import dataclass

from aws_lambda_powertools.utilities.idempotency import (
    DynamoDBPersistenceLayer, IdempotencyConfig, idempotent_function)

dynamodb = DynamoDBPersistenceLayer(table_name="idem")
config =  IdempotencyConfig(
    event_key_jmespath="messageId",  # see Choosing a payload subset section
    use_local_cache=True,
)

@dataclass
class OrderItem:
    sku: str
    description: str

@dataclass
class Order:
    item: OrderItem
    order_id: int


@idempotent_function(data_keyword_argument="order", config=config, persistence_store=dynamodb)
def process_order(order: Order):
    return f"processed order {order.order_id}"


order_item = OrderItem(sku="fake", description="sample")
order = Order(item=order_item, order_id="fake-id")

# `order` parameter must be called as a keyword argument to work
process_order(order=order)
```

### Pydantic support

```python
from aws_lambda_powertools.utilities.idempotency import (
    DynamoDBPersistenceLayer, IdempotencyConfig, idempotent_function)
from aws_lambda_powertools.utilities.parser import BaseModel

dynamodb = DynamoDBPersistenceLayer(table_name="idem")
config =  IdempotencyConfig(
    event_key_jmespath="messageId",  # see Choosing a payload subset section
    use_local_cache=True,
)


class OrderItem(BaseModel):
    sku: str
    description: str


class Order(BaseModel):
    item: OrderItem
    order_id: int


@idempotent_function(data_keyword_argument="order", config=config, persistence_store=dynamodb)
def process_order(order: Order):
    return f"processed order {order.order_id}"


order_item = OrderItem(sku="fake", description="sample")
order = Order(item=order_item, order_id="fake-id")

# `order` parameter must be called as a keyword argument to work
process_order(order=order)
```

**Checklist**

 

* [X] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [X] Update tests
* [x] Update docs
* [X] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 1]
```diff
@@ -226,7 +225,13 @@ def _generate_hash(self, data: Any) -> str:
             Hashed representation of the provided data
 
         """
-        data = getattr(data, "raw_event", data)  # could be a data class depending on decorator order
+        if hasattr(data, "__dataclass_fields__"):
```
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: @heitorlessa - not sure if there is a cleaner way of doing this without breaking python 3.6

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: Luckily 3.6 is EOL in 6 days ;-) so we will be able to drop it fully soon (need to check with the Lambda team).

In the meantime, can we move this serialisation into a shared function -or private method for now- as we will later support Pydantic models too?

We will be reuse this serialisation for Logger and possibly others too, e.g. be able to serialize Event Source 
Data Class, Dataclasses, and Pydantic Model consistently within Powertools

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: @heitorlessa - just as a dict or list? coz json serialization is not always consistent throughout

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: Not sure I follow? You mean how are we going to handle the return if it's a list?

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:228]: Thanks for clarifying on messaging, let's keep as a private method - we want the dict only. 

Later, we can turn into a reusable function to get the dict value then serialize to JSON str for the encoder

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py]: @heitorlessa - i have added  prepare method.

**Review by heitorlessa — CHANGES_REQUESTED** [review]
🤦 my comments were kept in pending... only one ask really - could you add a test with event_jmespath to use a given field of a dataclass?

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py]: @heitorlessa - this is based on the python 3 source code

https://github.com/python/cpython/blob/29ea68bd1dcf30842c2ed908a6d815bc1d90f484/Lib/dataclasses.py#L1210

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 2]
```diff
@@ -105,6 +104,21 @@ def response_json_as_dict(self) -> dict:
         return json.loads(self.response_data)
 
 
+def _prepare_data(data: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:107]: Yes. There is a "isdataclass" check in python 3.8

File: tests/functional/idempotency/test_idempotency.py [hunk 3]
```diff
@@ -1057,3 +1057,33 @@ def two(data):
...
     assert len(persistence_store.table.method_calls) == 4
+
+
+def test_idempotent_function_dataclasses():
+    try:
+        # Scenario idempotent_function should allow for python dataclasses
+        from dataclasses import asdict, dataclass
+
+        @dataclass
+        class Foo:
```
[the remark this answers is not in the document]
Comment (michaelbrewer) [reply tests/functional/idempotency/test_idempotency.py:1069]: @heitorlessa  - not sure what you mean. jmespath search on a data class will always return none.

[the remark this answers is not in the document]
Comment (heitorlessa) [reply tests/functional/idempotency/test_idempotency.py:1069]: I see. If we move this up, we could get the dict earlier, event_jmespath_key and jmespath validation would also work like any dict.

File: aws_lambda_powertools/utilities/idempotency/persistence/base.py [hunk 4]
```diff
@@ -226,7 +225,6 @@ def _generate_hash(self, data: Any) -> str:
             Hashed representation of the provided data
 
         """
-        data = getattr(data, "raw_event", data)  # could be a data class depending on decorator order
```
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/idempotency/persistence/base.py:229]: @heitorlessa - so doing it here is too late, so i am moved it up to the earliest possible point.

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/idempotency/persistence/base.py:229]: GitHub UI didn't allow me to make a comment upward, tks a lot for moving it upward - I'll check it

**Review by heitorlessa — APPROVED** [review]

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**